utopia-vscode-extension/src/utopia-fs.ts

```diff
@@ -193,127 +205,189 @@ export class UtopiaFSExtension
-    options: { create: boolean; overwrite: boolean },
-  ): Promise<void> {
-    const path = fromUtopiaURI(uri)
+  writeFile(uri: Uri, content: Uint8Array, options: { create: boolean; overwrite: boolean }) {
+    const path = uri.path
+    const fileExists = exists(path)
     if (!options.create || !options.overwrite) {
-      const fileExists = await exists(path)
       if (!fileExists && !options.create) {
         throw FileSystemError.FileNotFound(uri)
       } else if (fileExists && !options.overwrite) {
         throw FileSystemError.FileExists(uri)
       }
     }
 
-    await writeFileSavedContent(path, content)
+    writeFileSavedContent(path, content)
+    if (fileExists) {
+      this.notifyFileChanged(path)
+    } else {
+      this.notifyFileCreated(path)
+    }
+  }
+
+  ensureDirectoryExists(pathToEnsure: string) {
+    const allPaths = allPathsUpToPath(pathToEnsure)
+    let createdDirectories: Array<string> = []
+    for (const pathToCreate of allPaths) {
+      const existingNode = getItem(pathToCreate)
+      if (existingNode == null) {
+        createDirectoryWithoutError(pathToCreate)
+        createdDirectories.push(pathToCreate)
+      } else if (isFile(existingNode)) {
+        throw isNotDirectoryError(pathToCreate)
+      }
+    }
+
+    createdDirectories.forEach((createdDirectory) => this.notifyFileCreated(createdDirectory))
+  }
+
+  writeProjectFile(projectFile: ProjectFile) {
+    switch (projectFile.type) {
+      case 'PROJECT_DIRECTORY': {
+        const { filePath } = projectFile
+        this.ensureDirectoryExists(filePath)
+        break
+      }
+      case 'PROJECT_TEXT_FILE': {
+        const { filePath, savedContent, unsavedContent } = projectFile
+        const fileExists = exists(filePath)
+        const alreadyExistingFile = fileExists ? readFileAsUTF8(filePath) : null
+        const fileDiffers =
+          alreadyExistingFile == null ||
+          alreadyExistingFile.content !== savedContent ||
+          alreadyExistingFile.unsavedContent !== unsavedContent
+        if (fileDiffers) {
+          // Avoid pushing a file to the file system if the content hasn't changed.
+          writeFileAsUTF8(filePath, savedContent, unsavedContent)
+
+          if (fileExists) {
+            this.notifyFileChanged(filePath)
+          } else {
+            this.notifyFileCreated(filePath)
+          }
+        }
+        break
+      }
+      default:
+        const _exhaustiveCheck: never = projectFile
+        throw new Error(`Invalid file projectFile type ${projectFile}`)
+    }
+  }
+
+  delete(uri: Uri, options: { recursive: boolean }) {
+    this.silentDelete(uri.path, options)
+    commands.executeCommand('utopia.toUtopiaMessage', vsCodeFileDelete(uri.path))
   }
 
-  async delete(uri: Uri, options: { recursive: boolean }): Promise<void> {
-    const path = fromUtopiaURI(uri)
-    await deletePath(path, options.recursive)
+  silentDelete(path: string, options: { recursive: boolean }) {
```

Comment:
"silent" because it doesn't send the message to Utopia. It still emits the event. I couldn't think of a better name here tbh